File: SortAlgorithms.java
public class SortAlgorithms {

    private static final int INSERTION_SORT_THRESHOLD = 10; // Threshold for switching to insertion sort

    // Method to perform quick sort with hybrid approach
    public void quickSort(int[] arr) {
        quickSortIterative(arr, 0, arr.length - 1);
    }

    private void quickSortIterative(int[] arr, int low, int high) {
        int[] stack = new int[high - low + 1];
        int top = -1;
        stack[++top] = low;
        stack[++top] = high;

        while (top >= 0) {
            high = stack[top--];
            low = stack[top--];

            if (high - low < INSERTION_SORT_THRESHOLD) {
                insertionSort(arr, low, high);
                continue;
            }

            int pi = partition(arr, low, high);

            if (pi - 1 > low) {
                stack[++top] = low;
                stack[++top] = pi - 1;
            }

            if (pi + 1 < high) {
                stack[++top] = pi + 1;
                stack[++top] = high;
            }
        }
    }

    private int partition(int[] arr, int low, int high) {
        int pivot = arr[low + (high - low) / 2]; // Median-of-three or random pivot selection
        int i = low;
        int j = high;

        while (i <= j) {
            while (arr[i] < pivot) {
                i++;
            }
            while (arr[j] > pivot) {
                j--;
            }
            if (i <= j) {
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
                i++;
                j--;
            }
        }
        return i;
    }

    private void insertionSort(int[] arr, int low, int high) {
        for (int i = low + 1; i <= high; i++) {
            int key = arr[i];
            int j = i - 1;
            while (j >= low && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
    }

    // Iterative version of merge sort
    public void mergeSort(int[] arr) {
        int n = arr.length;
        int[] temp = new int[n];
        for (int currSize = 1; currSize < n; currSize *= 2) {
            for (int leftStart = 0; leftStart < n - 1; leftStart += 2 * currSize) {
                int mid = Math.min(leftStart + currSize - 1, n - 1);
                int rightEnd = Math.min(leftStart + 2 * currSize - 1, n - 1);
                merge(arr, temp, leftStart, mid, rightEnd);
            }
        }
    }

    private void merge(int[] arr, int[] temp, int leftStart, int mid, int rightEnd) {
        int i = leftStart, j = mid + 1, k = leftStart;

        while (i <= mid && j <= rightEnd) {
            if (arr[i] <= arr[j]) {
                temp[k++] = arr[i++];
            } else {
                temp[k++] = arr[j++];
            }
        }
        while (i <= mid) {
            temp[k++] = arr[i++];
        }
        while (j <= rightEnd) {
            temp[k++] = arr[j++];
        }
        for (i = leftStart; i <= rightEnd; i++) {
            arr[i] = temp[i];
        }
    }

    // Method to fill the array with sorted numbers
    public void fillArrayWithSortedNumbers(int[] array) {
        for (int i = 0; i < array.length; i++) {
            array[i] = i; // Fill array with sorted sequence [0, 1, 2, ..., array.length - 1]
        }
    }

    public static void main(String[] args) {
        int[] arraySizes = {1000, 5000, 10000, 25000, 50000, 75000, 100000, 1000000, 5000000, 10000000, 15000000, 20000000, 25000000, 30000000, 35000000, 40000000, 45000000, 50000000}; // Different input sizes

        SortAlgorithms obj = new SortAlgorithms();

        for (int size : arraySizes) {
            int[] sortedNumbers = new int[size];
            obj.fillArrayWithSortedNumbers(sortedNumbers);

            // Test Quick Sort
            int[] quickSortArray = sortedNumbers.clone();
            long start = System.nanoTime();
            obj.quickSort(quickSortArray);
            long end = System.nanoTime();
            long elapsedTimeInNanoseconds = end - start;
            double elapsedTimeInMilliseconds = elapsedTimeInNanoseconds / 1_000_000.0;
            System.out.println("Array size: " + size);
            System.out.println("Time taken for quick sort: " + elapsedTimeInMilliseconds + " milliseconds");

            // Test Merge Sort
            int[] mergeSortArray = sortedNumbers.clone();
            start = System.nanoTime();
            obj.mergeSort(mergeSortArray);
            end = System.nanoTime();
            elapsedTimeInNanoseconds = end - start;
            elapsedTimeInMilliseconds = elapsedTimeInNanoseconds / 1_000_000.0;
            System.out.println("Time taken for merge sort: " + elapsedTimeInMilliseconds + " milliseconds");
            System.out.println();
        }
    }
}

File: Python-Files/SortAlgorithms.py
class SortAlgorithms:
    INSERTION_SORT_THRESHOLD = 10

    def quick_sort(self, arr):
        self._quick_sort_iterative(arr, 0, len(arr) - 1)

    def _quick_sort_iterative(self, arr, low, high):
        stack = [(low, high)]
        while stack:
            low, high = stack.pop()
            if high - low < self.INSERTION_SORT_THRESHOLD:
                self._insertion_sort(arr, low, high)
                continue

            pivot_index = self._partition(arr, low, high)
            if pivot_index - 1 > low:
                stack.append((low, pivot_index - 1))
            if pivot_index < high:
                stack.append((pivot_index, high))

    def _partition(self, arr, low, high):
        pivot = arr[low + (high - low) // 2]
        i, j = low, high
        while i <= j:
            while arr[i] < pivot:
                i += 1
            while arr[j] > pivot:
                j -= 1
            if i <= j:
                arr[i], arr[j] = arr[j], arr[i]
                i += 1
                j -= 1
        return i

    def _insertion_sort(self, arr, low, high):
        for i in range(low + 1, high + 1):
            key = arr[i]
            j = i - 1
            while j >= low and arr[j] > key:
                arr[j + 1] = arr[j]
                j -= 1
            arr[j + 1] = key

    def merge_sort(self, arr):
        n = len(arr)
        temp = [0] * n
        size = 1
        while size < n:
            for left_start in range(0, n, 2 * size):
                mid = min(left_start + size - 1, n - 1)
                right_end = min(left_start + 2 * size - 1, n - 1)
                self._merge(arr, temp, left_start, mid, right_end)
            size *= 2

    def _merge(self, arr, temp, left_start, mid, right_end):
        i, j, k = left_start, mid + 1, left_start
        while i <= mid and j <= right_end:
            if arr[i] <= arr[j]:
                temp[k] = arr[i]
                i += 1
            else:
                temp[k] = arr[j]
                j += 1
            k += 1

        while i <= mid:
            temp[k] = arr[i]
            i += 1
            k += 1

        while j <= right_end:
            temp[k] = arr[j]
            j += 1
            k += 1

        for i in range(left_start, right_end + 1):
            arr[i] = temp[i]

    def generate_sorted_numbers(self, size):
        return list(range(size))


if __name__ == "__main__":
    array_sizes = [1000, 5000, 10000, 25000, 50000, 75000, 100000, 1000000, 5000000]
    sorter = SortAlgorithms()

    for size in array_sizes:
        sorted_array = sorter.generate_sorted_numbers(size)

        quick_sort_array = sorted_array[:]
        sorter.quick_sort(quick_sort_array)
        print(f"Array of size {size} sorted using Quick Sort.")

        merge_sort_array = sorted_array[:]
        sorter.merge_sort(merge_sort_array)
        print(f"Array of size {size} sorted using Merge Sort.")
